src/lazy/text/raw/v1_1/reader.rs

```diff
@@ -165,6 +166,119 @@ pub(crate) mod system_macros {
+        self.address
+    }
+}
+
+/// Types that may be able to be resolved to a macro ID.
+/// This is used by the writer to accept user-specified types to an ID based on the current encoding context.
+pub trait MacroIdLike<'a>: Sized {
+    fn as_macro_id_ref(&self) -> MacroIdRef<'a>;
+
+    fn prefer_name(&self) -> MacroIdRef<'a> {
+        // By default, change nothing.
+        self.as_macro_id_ref()
+    }
+
+    fn prefer_address(&self) -> MacroIdRef<'a> {
+        // By default, change nothing.
+        self.as_macro_id_ref()
+    }
+
+    fn resolve(self, macro_table: &'a MacroTable) -> IonResult<ResolvedId<'a>> {
+        use MacroIdRef::*;
+        let id = self.as_macro_id_ref();
+        match id {
+            LocalName(name) => {
+                if let Some(address) = macro_table.address_for_name(name) {
+                    return Ok(ResolvedId::new(Some(name), address));
+                };
+            }
+            LocalAddress(address) => {
+                if let Some(name) = macro_table
+                    .macro_at_address(address)
+                    .and_then(|macro_ref| macro_ref.name())
+                {
+                    return Ok(ResolvedId::new(Some(name), address));
+                }
+            }
+            SystemAddress(_system_address) => {
+                todo!("resolving qualified macro IDs")
+            }
+        };
+        IonResult::encoding_error(format!("could not find macro with ID {id:?}"))
+    }
+}
+
+impl<'a> MacroIdLike<'a> for &'a Macro {
+    fn as_macro_id_ref(&self) -> MacroIdRef<'a> {
+        MacroIdRef::LocalAddress(self.address())
+    }
+
+    fn prefer_name(&self) -> MacroIdRef<'a> {
+        match self.name() {
+            Some(name) => MacroIdRef::LocalName(name),
+            None => MacroIdRef::LocalAddress(self.address()),
+        }
+    }
+
+    fn prefer_address(&self) -> MacroIdRef<'a> {
+        MacroIdRef::LocalAddress(self.address())
+    }
+
+    fn resolve(self, _macro_table: &'a MacroTable) -> IonResult<ResolvedId<'a>> {
+        // TODO: Confirm that the macro at the saved address is the same one
+        Ok(ResolvedId::new(self.name(), self.address()))
+    }
+}
+
+impl<'a, T> MacroIdLike<'a> for T
+where
+    MacroIdRef<'a>: From<T>,
+    T: Copy,
+{
+    fn as_macro_id_ref(&self) -> MacroIdRef<'a> {
+        (*self).into()
+    }
+}
+
+impl<'a> MacroIdLike<'a> for ResolvedId<'a> {
+    fn as_macro_id_ref(&self) -> MacroIdRef<'a> {
+        self.prefer_name()
+    }
+
+    fn prefer_name(&self) -> MacroIdRef<'a> {
+        match self.name() {
+            Some(name) => MacroIdRef::LocalName(name),
+            None => MacroIdRef::LocalAddress(self.address()),
+        }
+    }
+
+    fn prefer_address(&self) -> MacroIdRef<'a> {
+        MacroIdRef::LocalAddress(self.address())
+    }
+
+    fn resolve(self, _macro_table: &'a MacroTable) -> IonResult<ResolvedId<'a>> {
+        // ResolvedId is already resolved
+        Ok(self)
+    }
```

Comment:
🪧 This is the important bit. If you pass an unresolved identifier to a raw eexp writer, it will try to resolve it. If you pass a resolved identifier to a raw eexp writer (say, from the managed writer that already resolved it), it won't need to resolve it again.